pkg/cmd/pulumi/state_repair.go

```diff
@@ -0,0 +1,289 @@
+	Stack     string
+	Colorizer colors.Colorization
+	Yes       bool
+}
+
+func newStateRepairCommand() *cobra.Command {
+	stateRepair := &stateRepairCmd{
+		Args: &stateRepairArgs{
+			Colorizer: cmdutil.GetGlobalColorization(),
+		},
+		Stdin:        os.Stdin,
+		Stdout:       os.Stdout,
+		Stderr:       os.Stderr,
+		Workspace:    pkgWorkspace.Instance,
+		LoginManager: DefaultLoginManager,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "repair",
+		Short: "Repair an invalid state",
+		Long: `Repair an invalid state,
+
+This command can be used to repair an invalid state file. It will attempt to
+sort resources that appear out of order and remove references to resources that
+are no longer present in the state. If the state is already valid, this command
+will not attempt to make or write any changes. If the state is not already
+valid, and remains invalid after repair has been attempted, this command will
+not write any changes.
+`,
+		Args: cmdutil.NoArgs,
+		Run: runCmdFunc(func(cmd *cobra.Command, args []string) error {
+			cmd.Flags().Visit(func(f *pflag.Flag) {
+				stateRepair.FlagsString += fmt.Sprintf(" --%s=%q", f.Name, f.Value)
+			})
+
+			ctx := cmd.Context()
+			err := stateRepair.run(ctx)
+
+			return err
+		}),
+	}
+
+	cmd.Flags().StringVarP(&stateRepair.Args.Stack,
+		"stack", "s", "", "The name of the stack to operate on. Defaults to the current stack")
+	cmd.Flags().BoolVarP(&stateRepair.Args.Yes,
+		"yes", "y", false, "Automatically approve and perform the repair")
+
+	return cmd
+}
+
+func (cmd *stateRepairCmd) run(ctx context.Context) error {
+	// We need to disable integrity checking since this command exists solely to repair invalid states and enabling
+	// integrity checking would prevent us from even loading them in the first place. Currently integrity checking is
+	// controlled by a hacky global variable, so this is the best we can do, but hopefully it and consequently this code
+	// can be refactored.
+	backend.DisableIntegrityChecking = true
+	defer func() { backend.DisableIntegrityChecking = false }()
+
+	displayOpts := display.Options{
+		Color: cmd.Args.Colorizer,
+	}
+	s, err := requireStack(
+		ctx,
+		cmd.Workspace,
+		cmd.LoginManager,
+		cmd.Args.Stack,
+		stackOfferNew,
+		displayOpts,
+	)
+	if err != nil {
+		return err
+	}
+
+	snap, err := s.Snapshot(ctx, stack.DefaultSecretsProvider)
+	if err != nil {
+		return err
+	} else if snap == nil {
+		return nil
+	}
+
+	sink := diag.DefaultSink(cmd.Stdout, cmd.Stderr, diag.FormatOptions{
+		Color: cmd.Args.Colorizer,
+	})
+
+	// If the snapshot is already valid, we won't touch it.
+	initialErr := snap.VerifyIntegrity()
+	if initialErr == nil {
+		sink.Infof(diag.RawMessage("" /*urn*/, "The snapshot is already valid; skipping repair"))
+		return nil
+	}
+
+	// The snapshot isn't valid, so we should proceed. If the user has passed the --yes flag, we'll skip the confirmation
+	// prompt and proceed directly to the repair.
+	if !cmd.Args.Yes {
```

Comment:
Would it be possible to actually show the user what we would change, so they could make a more informed decision on what they are confirming?  E.g. we could potentially verify the changed stack (no need to confirm if the verification fails) and show them a diff between the old and the new?

Reply:
I've had a crack at this and we now render:

* What will be reordered
* What will be modified

It's not perfect since things could be reordered *and* modified (and I've not come up with a neat way to combine results of sorting and pruning just yet), but I feel like this is a good start. Thoughts?

![image](https://github.com/user-attachments/assets/d508363a-5555-4483-878b-eeda1ec90cd9)
![image](https://github.com/user-attachments/assets/43fd615e-fc57-45d9-9335-62e372a623de)